### DISCUSSION TITLE
Please use complete URL for stats

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#177] [Anything Else]
"My Blogsoftware uses the part after the question mark to differentiate the different blog articles.

Example: https://www.deimeke.net/dirk/blog/index.php?/archives/4015-Lesen!.html

The stats only show “/dirk/blog/index.php”, which lacks the important information."

**Update:** We now have a way to track custom query parameters and get the complete page URLs in your Top Pages report. See the instructions https://plausible.io/docs/custom-query-params

**Comment by ddeimeke** [comment]
I will be happy to try plausible again once this feature is implemented.

**Comment by tckb** [comment]
@ddeimeke the example that you've shared _does not_ conform to [standard URI scheme ](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Generic_syntax) for identifying resources. 

A standard syntax is of the following format: 
```
http://[baseurl]/[path][?query]
```
the important information you seek is in the "query" part of the URL and the tracking is possible for different paths which does not include queries.

**Comment by ddeimeke** [comment]
@tckb It depends on your point of view.

You query the blog for the article called "/archives/4015-Lesen!.html".

I understand, that it is not possible (yet), but nevertheless it is how my Blogsoftware delivers the articles. If plausible.io does not recognize this, it is the wrong software for me.

**Comment by ukutaht** [comment]
By default we strip all query parameters except attribution related ones (`ref`, `utm_source`, etc.). I did it this way because:
* Query params often contain personal or sensitive data
* Query params are rarely used to distinguish different pages (as your blog software does)

I would like to keep stripping query params by default but we should probably allow a configurable whitelist so people can choose which query params to send to the server.

**Comment by onesork** [comment]
Hi @ukutaht, 

I run into issues due to this limitation as well.

> Query params often contain personal or sensitive data

I don't think that's frequent. Google Analytics has a policy that says you cannot store Personally Identifiable Information (PII) anywhere within GA (incl. URLs). Yet, you rarely need to make any adjustments to comply with this policy.

Of course, it depends on your definition of sensitive data. Most common would probably be queries from an internal search that are frequently stored in the URL. Are these considered sensitive as well?

> Query params are rarely used to distinguish different pages (as your blog software does)

I'm afraid that's unfortunately not on par with everyday reality. Many websites use parameters for filters/labels/categories that you want to be able to track. Others use that to communicate some form of state that's handy for conversion tracking.

For example, Ghost's default conversion goal destination is /?action=subscribe&success=true - which I'm not able to track easily (to give one example).

But I guess the whitelisting would solve most of these issues.

Thanks!

Best,

Jan

**Comment by websocket98765** [comment]
Or instead of a whitelist, maybe offer a setting to enable or disable query string tracking for all or some URLs, and allow some URLs to be deleted directly from analytics in case PII gets through. It sounds like a headache to maintain a whitelist for all routes on my site, but maybe it'd be ok.

**Comment by bkazez** [comment]
This is essential to help us track search requests.

**Comment by nico-leb** [comment]
Hi,

I also think a whitelist setting would be something usefull, at least for Page View goals. 

I often have to deal with "?success=ok" "order_confirm=xxxx" query parameters to track something.

Thanks!

**Comment by Alpine418** [comment]
Please add support for query params. I want to track search and filter form requests. Currently this is not possible. 

Whitelist looks like a good idea. Any plans when this got implemented?

**Reply by metmarkosaric** [reply]
We cannot make any promises on this. We are a small team with limited development resources so need to focus on things one thing at a time.

**Comment by CTFries** [comment]
+1 for the need for query parameters. Running a small news site(500-1m view/month) and our articles are identified via url param. 

Would love to make this our primary analytics tool, but for now it might just be covering a gap if we can't get params

**Comment by PeteKraft** [comment]
Just adding my vote.
I just signed up for the trial, and first impression was very good...until I noticed URL parameters are being stripped.

I currently need to implement analytics for a client who wants to track which of their news articles get the most clicks... and surely the news articles are seperated on the basis of URL params (newsitem?id=xxx)

**Reply by Alpine418** [reply]
Thats not an issue of the proxy. Plausible itself stripped the GET params. 

Read more here please: https://github.com/plausible/analytics/discussions/177

Looks like you already posted it in there. So why did you write it in here?

**Reply by PeteKraft** [reply]
???? who is talking about a proxy ?
And yes, I posted here...  I didn't mention proxy anywhere ???

**Reply by Alpine418** [reply]
Holy cow. I am so sorry. My fault. I tought I was in a other thread.

**Reply by PeteKraft** [reply]
I'll just add a quick note that I am approaching this a little different now... 
As suggested higher in this thread, I am now doing rewriterules inside the .htaccess file to create search engine friendly urls.
It's a bit much to wrap my head around, but it's coming toghether quite nice, and I can now see different news articles pop up in the analytics.

Still, I think a little bit more  granularity might be a nice option to have wether to allow for url parameters or not.

**Comment by mizuber** [comment]
Is here any update? The request seems quite old. Still not supported?

**Reply by metmarkosaric** [reply]
hi @mizuber! the work on this hasn't started yet and there are no short term plans to introduce it. we are a small team with limited development resources so need to focus on most frequently requested features first.

**Comment by TheCodeTherapy** [comment]
@metmarkosaric Hi there Marko!
While I understand the reasoning behind your decision to strip query parameters, I'd like to reinforce that in a lot of cases (like countless projects involving React single-page applications) in which query parameters control basic functionalities of the app (not providing any information about the user). I'll use my own personal WebGL portfolio as an example:

[https://mgz.me/?scene=TurbulentLight](https://mgz.me/?scene=TurbulentLight) Leads you to my website, with a specific DemoScene on my main renderer (as background)

[https://mgz.me/?scene=My90sRoom](https://mgz.me/?scene=My90sRoom) Leads you to a completely different scene

And I have, so far, 42 DemoScenes in this personal project. Several of my clients also adopt this way of choosing parameters for their apps that has nothing to do with user private information.

I loved Plausible, and this project would be absolutely perfect for me, but because I have my generative art exposed on third-party galleries, without the possibility of configuring the ?scene= query param as something that should be taken into consideration, I'm incapable of using Plausible to evaluate which one of my scenes are the most shared/accessed.

Obviously, there are route-based workarounds so I can have /path entry points, but for the sake of flexibility, I'd honestly ask you to reconsider this feature for the near future as I presume it would be very valuable for countless use-cases like mine.

Kindest regards (and thank you / kudos for this amazing project),

Marco.

**Reply by metmarkosaric** [reply]
thanks for the kind words and the feedback @TheCodeTherapy! makes sense. we haven't gotten around to this one yet unfortunately. we are a small team with limited development resources so need to focus on most frequently requested features first.

**Reply by bkazez** [reply]
I understand this logic and often apply it to my own software, but in this case you should also consider people who don't consider Plausible, or leave, because it doesn't have this. It is considerably less useful without this.

Is there a way the public can see your feature request list in order of frequency, to have a sense of what is higher priority?

**Reply by metmarkosaric** [reply]
we look at all the public feedback here on github and in social media. we also look at all the private feedback we get via email. and then we mix this with our own opinion on what we want Plausible to be. as explained earlier in this thread, we plan to keep stripping query params by default as we're a privacy first tool but we may allow a configurable whitelist so people can choose which query params to send to the server. this is a bit more of a niche feature and hasn't been frequently requested so we have no short term plans that we can tell you about. 

currently the main focus is on adding cities / states / regions as it's by far the most requested feature. after that, we will probably focus on full export / import which is another one that many people ask for. we will post in here when there are more concrete plans for this feature.

**Comment by merkoski** [comment]
Bump. I just signed up for Plausible today and this would be amazing if it worked to see the query portion of a url .... without it, not much value and I probably won't pay beyond the free trial.

**Reply by ukutaht** [reply]
Thanks @merkoski.  I realize this is a showstopper for those who need it. The amount of comments here definitely lifts this up in our priority list. It seems this is one of the top issues we have so it's just a matter of time until it gets fixed. I have a lot on my plate so can't make any promises as to when.

We're always open for PRs.

**Reply by metmarkosaric** [reply]
this could be done as another [script extension](https://plausible.io/docs/script-extensions) perhaps?

**Reply by ukutaht** [reply]
Yeah I think so

**Comment by gematria47** [comment]
Guys your platform is great, the reporting is so clean. Like stripe for web traffic, cutting through the muck. Unfortunately URL Parameters are central to our tracking. Would love to see this feature built someday.

**Comment by sdoncho** [comment]
@ukutaht
I am a contractor that rotates contracts every 3 months or so so I see quite a lot of different commercial client setups. I can tell you from what I've seen half my clients don't want to measure their natural users and where they come from. They are just too small to rank organically anyway.  What analytics is mostly used for is they set up different google ads with query parameters (e.g., ?ad=shrek) and try to measure the conversion rate of each ad. So in my (most likely very skewed sample) eyes, this otherwise amazing product does not support the most widely used analytics use case I've seen. 

But I REALY want to migrate away from google ads. Maybe a workaround for people interested: You can potentially extract the query parameters and create a custom event with properties which you can filter for?

**Reply by ukutaht** [reply]
Thanks @sdoncho. This is requested so heavily that I think I'll have no choice but to work on this soon :)

There's this proposal for another feature where you can manually transform the URL on the client side as well: https://github.com/plausible/analytics/discussions/136#discussioncomment-1402561

It wouldn't be super streamlined but it would be extremely flexible and allow you to parse out query params on your end before sending the URL. What do you think?

**Reply by ukutaht** [reply]
>  allow you to parse out query params on your end before sending the URL. What do you think?

Scratch that. Query params are stripped on the server side so it has to be a backend change.

I have a design in mind for allowing custom query params but honestly it won't be until early next year that I can tackle it. Sorry.

> But I REALY want to migrate away from google ads. Maybe a workaround for people interested: You can potentially extract the query parameters and create a custom event with properties which you can filter for?

Yeah, you could basically send them as custom props to a custom event as a workaround for now.

**Reply by metmarkosaric** [reply]
> What analytics is mostly used for is they set up different google ads with query parameters (e.g., ?ad=shrek) and try to measure the conversion rate of each ad. So in my (most likely very skewed sample) eyes, this otherwise amazing product does not support the most widely used analytics use case I've seen.

thanks @sdoncho! just a note that we already do support all the most widely used marketing parameters ref, source, utm_source, utm_medium and utm_campaign. see https://plausible.io/docs/manual-link-tagging

**Comment by Haestus** [comment]
Came here to confirm this is absolutely needed. My custom site uses query parameters for different pages.

Using query parameters is a completely valid way to show different pages, so I'm a little surprised this isn't in yet :)

**Comment by michaelwschultz** [comment]
Whew, glad to hear this is going to be worked on soon. Thought I just wasted a bunch of time contributing to https://github.com/4lejandrito/next-plausible to support the localhost extension.

Probably going to go with the custom event work around proposed by @sdoncho for now. 👍

**Reply by michaelwschultz** [reply]
Looks like I'll need to add the manual extension as well.

**Comment by fili** [comment]
I am surprised nobody mentioned pagination or searches yet as commonly used with query parameters, like ?page=44 or ?p=2, or ?s=test or ?query=123 or ?q=234 

So many use-cases for query parameters without tracking or PII data, filtering was indeed already mentioned for ecommerce, that query parameters need to be tracked always. Just my 2cts.

**Comment by Haestus** [comment]
Any news on this?

I'm a paying customer, but if this doesn't get solved (soon) you're forcing me to look elsewhere.

**Reply by metmarkosaric** [reply]
Hi @Haestus! There's no news on this and no short term plans that I can tell you about. We are focused on the more frequently requested features such as export/import at the moment. We are a small team with limited development resources so need to focus on most frequently requested features first. Thanks for understanding!

**Reply by Haestus** [reply]
Thank you for the honest answer

**Comment by metmarkosaric** [comment]
We now have a way to track custom query parameters and get the complete page URLs in your Top Pages report. See the instructions https://plausible.io/docs/custom-query-params

**Reply by fili** [reply]
Can this method be used to track fragments as well?

As discussed in https://github.com/plausible/analytics/discussions/835

**Reply by ukutaht** [reply]
A similar approach would work but the `prepareUrl` function needs to be different to track a fragment

**Comment by kimATwork** [comment]
I'v been mailing with  @RobertJoonas so no need to respond. I'm evaluating Plausible Cloud for use by our customers, and the lack of query params is unfortunately a showstopper for most. We did work around this with custom-query-params but really need GA imports to follow suite. Plausible was on top of our short list partly for its attractive OSS-approach, so a 'proper' query params support would bring it back.  At this point lots of sites are looking for GA alternatives, making this a closing window of opportunity as they choose their next analytics solutions. I'm sure you're aware, keep up the good work!

**Reply by ukutaht** [reply]
Thanks @kimATwork!

Would you prefer all query params to be collected automatically and the whole URL used to separate the pages out? I think this is how GA does it.

We've thought about adding query param support based on an allowlist that you need to configure. Which way would make more sense to you?

**Reply by kimATwork** [reply]
Thanks for responding!
From the use case i have at hand I'd say either the full URL or an allowlist.
An allowlist definitely makes sense, and I see a way to let relevant stakeholders approve the transition to Plausible. I.e. we could do a GA import based on an allowlist, with a subsequent approval. If the approval process reveals missing parameters, we can just delete the GA data and redo the import.

For new data we got around the current limitation using script.manual.js so the posted URLs looks like "artikkel.html¿tid=212215&sek=113027". This is not optimal and we would prefer a functional URL that allows '?'. So the current workaround with script.manual.js and allowing '?' would be great.

Even though current GA policy requires PII to be removed, there is an obvious possibility that both historic GA data and new data collected by Plausible can contain PII. I think the allowlist approach in both cases encourages the evaluation of parameters relevance, and by that reduces the risk of breaking GDPR compliance by accident. There will be a risk of lost data if any new parameters are introduced later on without updating the allowlist. But I think that's acceptable and a good compromise.

**Reply by ukutaht** [reply]
>  I think the allowlist approach in both cases encourages the evaluation of parameters relevance, and by that reduces the risk of breaking GDPR compliance by accident. There will be a risk of lost data if any new parameters are introduced later on without updating the allowlist. But I think that's acceptable and a good compromise.

Thanks, this is valuable feedback.

I think there are even more reasons to make it based on an allowlist as well. There are many query params that don't signify a different page or screen. Take a search result page for example like `https://mysite.com/search?q=my+search+query`. Should this show up as a separate page entry in Plausible from `https://mysite.com/search?q=different+search+query`? Probably not, it should show `/search` as a single entry with `q` as a custom dimension or something like that.

So you see it's not so easy make something that immediately works with no extra work from the user. But I think we've collected enough feedback and can design a good solution for this issue.

**Comment by brandonburkett** [comment]
I see the allow list support in the docs, but what is the recommended approach for array params?  EX: `/search?type[]=1&type[]=2&includes[]=3&includes=4&page=2`

In our use case, pushState is triggered on each filter or next page.... and user can use browser forward / back without issue... yet since the path is the same, they are not (automatically) tracked as separate pageviews.  The allow list approach starts to loose context if they are converted to folder paths in plausible generically.  I understand I can customize the `prepareUrl` function do whatever we need, but with multiple filter views, it could mean a lot of page-specific customization.

It would be great if there was a checkbox to allow for query params and to track pageviews for same path + diff query params.
